### PR TITLE
Solution of exercise two related to OOP. 

### PR DESCRIPTION
**Pull request by crisgahur**
Uploading exercise of ticket [BACK-11](https://vulky.atlassian.net/browse/BACK-11).

This is a POO exercise developed in Java. In this project you will find these concepts applied:

- Class.
- Methods
- Attributes
- Inheritance
- Extends.
- Implements
- Polymorphism.
- Interface
- Objects as parameter
- Model class
- Encapsulation

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,44 @@
-# OOPExerciseTwo
\ No newline at end of file
+# OOP Exercise TWO
+👋 Hi !This is a POO exercise developed in Java. In this proyect you will find these concepts applied:
```
Comment (MauroDecodificado) [inline README.md:2]: Love it ❤️

File: README.md [hunk 2]
```diff
@@ -1 +1,44 @@
...
+
+## IDE
+The workshop was developed with Netbeans IDE 16.
+
+## Status 
+In review for the manager. Last update 27/01/2023. No bugs detected.
+
+## Authors.
+Cristian Gallego.
+
```
Comment (MauroDecodificado) [inline README.md]: Delete empty space

File: src/main/java/com/mycompany/exercisetwo/AnimalModel.java [hunk 3]
```diff
@@ -0,0 +1,14 @@
+package com.mycompany.exercisetwo;
+
+/**
+ *
+ * @author Cristian Gallego
+ */
+public class AnimalModel {
```
Comment (MauroDecodificado) [inline src/main/java/com/mycompany/exercisetwo/AnimalModel.java:7]: Excercise is requesting at least 5 parameters, you just have 4

File: src/main/java/com/mycompany/exercisetwo/Death.java [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+
+import java.time.LocalDate;
+
+/**
+ *
+ * @author Cristian Gallego
+ */
+public interface Death {
+
+    public String die(LocalDate born, LocalDate die, String deathCause);
```
Comment (MauroDecodificado) [inline src/main/java/com/mycompany/exercisetwo/Death.java:11]: The exercise asked for it but I like the fact you use LocalDate for years, so I'm happy with this approach 😄

File: src/main/java/com/mycompany/exercisetwo/Humming.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    public void sleep() {
+        System.out.println("The humming usually sleep few hours");
+    }
+
+    @Override
+    public String die(LocalDate born, LocalDate die, String deathCause) {
+
+        Period period = Period.between(born, die);
+
+        return "The date of it's die is: " + die + ", lived: " + period.getYears() + " years, and the death     cause was: " + deathCause;
```
Comment (MauroDecodificado) [inline src/main/java/com/mycompany/exercisetwo/Humming.java]: What is that whole space between "and the death      cause was"?

File: src/main/java/com/mycompany/exercisetwo/Humming.java [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    @Override
+    public void sleep() {
+        System.out.println("The humming usually sleep few hours");
+    }
+
+    @Override
+    public String die(LocalDate born, LocalDate die, String deathCause) {
+
+        Period period = Period.between(born, die);
+
```
Comment (MauroDecodificado) [inline src/main/java/com/mycompany/exercisetwo/Humming.java]: Delete this empty space

File: src/main/java/com/mycompany/exercisetwo/Humming.java [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+    @Override
+    public void sleep() {
+        System.out.println("The humming usually sleep few hours");
+    }
+
+    @Override
+    public String die(LocalDate born, LocalDate die, String deathCause) {
+
+        Period period = Period.between(born, die);
```
Comment (MauroDecodificado) [inline src/main/java/com/mycompany/exercisetwo/Humming.java:22]: Awesome 👏

File: src/main/java/com/mycompany/exercisetwo/Humming.java [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+        return "The date of it's die is: " + die + ", lived: " + period.getYears() + " years, and the death     cause was: " + deathCause;
+
+    }
+
+    @Override
+    public String reincarnate(String couldReincarnate) {
+        return couldReincarnate;
+    }
+
+    public AnimalModel globalVariableAnimalModel;
```
Comment (MauroDecodificado) [inline src/main/java/com/mycompany/exercisetwo/Humming.java]: For good practices all global attributes must be at the top of the class

File: src/main/java/com/mycompany/exercisetwo/Humming.java [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+
+    public Humming() {
+
```
Comment (MauroDecodificado) [inline src/main/java/com/mycompany/exercisetwo/Humming.java]: Delete empty space

File: src/main/java/com/mycompany/exercisetwo/Humming.java [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+        AnimalModel x = new AnimalModel();
```
Comment (MauroDecodificado) [inline src/main/java/com/mycompany/exercisetwo/Humming.java]: Please do not use "x" developer do not know what is that, so please use a more meaningful name

File: src/main/java/com/mycompany/exercisetwo/Humming.java [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+
```
Comment (MauroDecodificado) [inline src/main/java/com/mycompany/exercisetwo/Humming.java]: Delete empty space

File: src/main/java/com/mycompany/exercisetwo/Humming.java [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+        globalVariableAnimalModel = x;
+
```
Comment (MauroDecodificado) [inline src/main/java/com/mycompany/exercisetwo/Humming.java]: Delete empty space

File: src/main/java/com/mycompany/exercisetwo/Humming.java [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+    }
+
+    public Humming(AnimalModel globalVariableAnimalModel) {
+
```
Comment (MauroDecodificado) [inline src/main/java/com/mycompany/exercisetwo/Humming.java]: Delete empty space

File: src/main/java/com/mycompany/exercisetwo/Humming.java [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+        globalVariableAnimalModel.color = "verde";
+        globalVariableAnimalModel.isDomesticable = false;
+        globalVariableAnimalModel.environment = "tropical";
+        globalVariableAnimalModel.kind = "bird";
+
```
Comment (MauroDecodificado) [inline src/main/java/com/mycompany/exercisetwo/Humming.java:45]: Delete empty space

File: src/main/java/com/mycompany/exercisetwo/Humming.java [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+    public Humming() {
+
+        AnimalModel x = new AnimalModel();
+
+        globalVariableAnimalModel = x;
+
+    }
+
+    public Humming(AnimalModel globalVariableAnimalModel) {
```
Comment (MauroDecodificado) [inline src/main/java/com/mycompany/exercisetwo/Humming.java]: This is not a global variable called this as `localVariableAnimalModel`

File: src/main/java/com/mycompany/exercisetwo/Humming.java [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+
+        globalVariableAnimalModel.color = "verde";
```
Comment (MauroDecodificado) [inline src/main/java/com/mycompany/exercisetwo/Humming.java]: If you have received by param the model, that means it comes with the information so you don't have to fulfill it just assign the local to global

File: src/main/java/com/mycompany/exercisetwo/Humming.java (hunk 10 above)
Comment (MauroDecodificado) [inline src/main/java/com/mycompany/exercisetwo/Humming.java]: You create the instance animal model and then immediately you assign it to the global so what information does this "x" have?

File: src/main/java/com/mycompany/exercisetwo/Cow.java [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+ *
+ * @author Cristian Gallego
+ */
+public class Cow extends Mammalian implements Death {
+
+    public String result;
+
+    @Override
+    public String die(LocalDate born, LocalDate die, String deathCause) {
+
```
Comment (MauroDecodificado) [inline src/main/java/com/mycompany/exercisetwo/Cow.java]: Delete empty space

File: src/main/java/com/mycompany/exercisetwo/Cow.java [hunk 18]
```diff
@@ -0,0 +1,47 @@
...
+ */
+public class Cow extends Mammalian implements Death {
+
+    public String result;
+
+    @Override
+    public String die(LocalDate born, LocalDate die, String deathCause) {
+        
+        Period period = Period.between(born, die);
+
```
Comment (MauroDecodificado) [inline src/main/java/com/mycompany/exercisetwo/Cow.java:15]: Delete empty space

File: src/main/java/com/mycompany/exercisetwo/Cow.java [hunk 19]
```diff
@@ -0,0 +1,47 @@
...
+import java.time.LocalDate;
+import java.time.Period;
+
+/**
+ *
+ * @author Cristian Gallego
+ */
+public class Cow extends Mammalian implements Death {
+
+    public String result;
```
Comment (MauroDecodificado) [inline src/main/java/com/mycompany/exercisetwo/Cow.java]: You do not have to keep this variable as global reference you can create it inside die method

File: src/main/java/com/mycompany/exercisetwo/Cow.java [hunk 20]
```diff
@@ -0,0 +1,47 @@
...
+        
+        Period period = Period.between(born, die);
+
+        if (period.getYears() % 2 == 0) {
+
+            result = "pair number";
+        } else {
+            result = "odd number";
+        }
+        return "The date of death was: " + die + ", the number is " + result + ", lived: " + period.getYears() + " years, and the death cause was: " + deathCause;
```
Comment (MauroDecodificado) [inline src/main/java/com/mycompany/exercisetwo/Cow.java:18]: You can improve this logic in "result" by using something called ternary operator

https://www.baeldung.com/java-ternary-operator
https://www.geeksforgeeks.org/java-ternary-operator-with-examples/

Comment (MauroDecodificado) [reply src/main/java/com/mycompany/exercisetwo/Cow.java:18]: Also remember the exercise says "Morir, debe calcular la edad y **multiplicarla por dos"**

File: src/main/java/com/mycompany/exercisetwo/Cow.java [hunk 21]
```diff
@@ -0,0 +1,47 @@
...
+
+    @Override
+    public void drinkMilk() {
+        System.out.println("The cow offer milk to the humans");
+    }
+
+    public AnimalModel globalVariableAnimalModel;
+
+    public Cow() {
+        AnimalModel x = new AnimalModel();
```
Comment (MauroDecodificado) [inline src/main/java/com/mycompany/exercisetwo/Cow.java]: Please do not use "x" developer do not know what is that, so please use a more meaningful name

File: src/main/java/com/mycompany/exercisetwo/Cow.java [hunk 22]
```diff
@@ -0,0 +1,47 @@
...
+    public String reincarnate(String couldReincarnate) {
+        return couldReincarnate + ", you can do it!";
+    }
+
+    @Override
+    public void drinkMilk() {
+        System.out.println("The cow offer milk to the humans");
+    }
+
+    public AnimalModel globalVariableAnimalModel;
```
Comment (MauroDecodificado) [inline src/main/java/com/mycompany/exercisetwo/Cow.java]: For good practices all global attributes must be at the top of the class

File: src/main/java/com/mycompany/exercisetwo/Cow.java [hunk 23]
```diff
@@ -0,0 +1,47 @@
...
+
+    public Cow() {
+        AnimalModel x = new AnimalModel();
+
```
Comment (MauroDecodificado) [inline src/main/java/com/mycompany/exercisetwo/Cow.java]: You are creating the object and new instance but does not have the information so please fulfill the information of the object (The attributes)

File: src/main/java/com/mycompany/exercisetwo/Cow.java [hunk 24]
```diff
@@ -0,0 +1,47 @@
...
+    public String result;
+
+    @Override
+    public String die(LocalDate born, LocalDate die, String deathCause) {
+        
+        Period period = Period.between(born, die);
+
+        if (period.getYears() % 2 == 0) {
+
```
Comment (MauroDecodificado) [inline src/main/java/com/mycompany/exercisetwo/Cow.java:17]: Delete empty space

**Review by MauroDecodificado — CHANGES_REQUESTED** [review]
Check commets

File: README.md (hunk 2 above)
Comment (crisgahur) [reply README.md]: Done

File: src/main/java/com/mycompany/exercisetwo/AnimalModel.java (hunk 3 above)
Comment (crisgahur) [reply src/main/java/com/mycompany/exercisetwo/AnimalModel.java:7]: Corrected

File: src/main/java/com/mycompany/exercisetwo/Humming.java (hunk 5 above)
Comment (crisgahur) [reply src/main/java/com/mycompany/exercisetwo/Humming.java]: My bad, corrected.

File: src/main/java/com/mycompany/exercisetwo/Humming.java (hunk 11 above)
Comment (crisgahur) [reply src/main/java/com/mycompany/exercisetwo/Humming.java]: Done

File: src/main/java/com/mycompany/exercisetwo/Humming.java (hunk 9 above)
Comment (crisgahur) [reply src/main/java/com/mycompany/exercisetwo/Humming.java]: Done

File: src/main/java/com/mycompany/exercisetwo/Humming.java (hunk 12 above)
Comment (crisgahur) [reply src/main/java/com/mycompany/exercisetwo/Humming.java]: Done

File: src/main/java/com/mycompany/exercisetwo/Humming.java (hunk 8 above)
Comment (crisgahur) [reply src/main/java/com/mycompany/exercisetwo/Humming.java]: Done

File: src/main/java/com/mycompany/exercisetwo/Humming.java (hunk 10 above)
Comment (crisgahur) [reply src/main/java/com/mycompany/exercisetwo/Humming.java]: Corrected

File: src/main/java/com/mycompany/exercisetwo/Cow.java (hunk 19 above)
Comment (crisgahur) [reply src/main/java/com/mycompany/exercisetwo/Cow.java]: Got it. Changed to a local scope.

File: src/main/java/com/mycompany/exercisetwo/Cow.java (hunk 17 above)
Comment (crisgahur) [reply src/main/java/com/mycompany/exercisetwo/Cow.java]: Done

File: src/main/java/com/mycompany/exercisetwo/Cow.java (hunk 24 above)
Comment (crisgahur) [reply src/main/java/com/mycompany/exercisetwo/Cow.java:17]: Done

File: src/main/java/com/mycompany/exercisetwo/Cow.java (hunk 20 above)
Comment (crisgahur) [reply src/main/java/com/mycompany/exercisetwo/Cow.java:18]: Done, ternary operator used and calculate corrected.

File: src/main/java/com/mycompany/exercisetwo/Cow.java (hunk 22 above)
Comment (crisgahur) [reply src/main/java/com/mycompany/exercisetwo/Cow.java]: Done

File: src/main/java/com/mycompany/exercisetwo/Cow.java (hunk 21 above)
Comment (crisgahur) [reply src/main/java/com/mycompany/exercisetwo/Cow.java]: Corrected

File: src/main/java/com/mycompany/exercisetwo/Cow.java (hunk 23 above)
Comment (crisgahur) [reply src/main/java/com/mycompany/exercisetwo/Cow.java]: Done

File: src/main/java/com/mycompany/exercisetwo/Humming.java (hunk 6 above)
Comment (crisgahur) [reply src/main/java/com/mycompany/exercisetwo/Humming.java]: Done

File: src/main/java/com/mycompany/exercisetwo/Humming.java (hunk 10 above)
Comment (crisgahur) [reply src/main/java/com/mycompany/exercisetwo/Humming.java]: Fulfill with the attributes information

File: src/main/java/com/mycompany/exercisetwo/Humming.java (hunk 15 above)
Comment (crisgahur) [reply src/main/java/com/mycompany/exercisetwo/Humming.java]: Corrected

File: src/main/java/com/mycompany/exercisetwo/Humming.java (hunk 14 above)
Comment (crisgahur) [reply src/main/java/com/mycompany/exercisetwo/Humming.java:45]: Done

File: src/main/java/com/mycompany/exercisetwo/AnimalModel.java [hunk 25]
```diff
@@ -10,5 +10,6 @@ public class AnimalModel {
     public String color;
```
Comment (MauroDecodificado) [inline src/main/java/com/mycompany/exercisetwo/AnimalModel.java]: Please create encapsulation on this class

**Review by MauroDecodificado — CHANGES_REQUESTED** [review]
Just one comment

File: src/main/java/com/mycompany/exercisetwo/AnimalModel.java (hunk 25 above)
Comment (crisgahur) [reply src/main/java/com/mycompany/exercisetwo/AnimalModel.java]: Done.

**Review by MauroDecodificado — APPROVED** [review]
PR is approved and now can be merged!

**Pull request merged**